### ISSUE TITLE
Замечания по BCG

### ISSUE THREAD

**Issue by dzheltkov** [3Jlou4eJluk/learning_stuff#1]
1. От матрицы требуются только размеры и матвек (на матрицу и транспонированную), поэтому она должна передаваться в функцию, осуществляющую метод BCG, или как размер и функция c параметром, умножение на матрицу или на транспонированную  (лучше использовать `std::functional`, тогда возможно передать и класс с реализованным `operator()` ), или как шаблон, в котом реализованы операции `matvec` и, например, `order`.
2. С++ обёртки для операций неэффективны.

**Comment by dzheltkov** [comment]
OpenMP версия зачтена.
В MPI версии лучше сделать, чтобы матрица хранилась распределённо (сейчас, насколько я понимаю, на каждом процессе хранится копия матрицы) и не в виде указателя на указатель. Кроме того, неплдохо бы выделить необходимые операции в функции (или погрузить в класс), чтобы код лучше читался.
И не хватае CMake правил для сборки.

**Comment by 3Jlou4eJluk** [comment]
Нет, в mpi версии хранение матрицы А происходит по строкам, вся
матрица(целиком) ни в одном из процессов не хранится. Т.е 1 строка матрица
в 0 процессе, 2 строка в 1 процессе, 3 строка в 0 процессе, 4 строка в 1
процессе(если у нас всего 2 процесса)

чт, 25 мая 2023 г. в 11:10, dzheltkov ***@***.***>:

**Comment by dzheltkov** [comment]
Да, неправильно понял суть выражений вида `if(my_rank == i%size)`.
В любом случае, таких выражений в целом быть не должно, разве только при загрузке матрицы.

Зачем в целом так распределять матрицу, почему не подряд идущими блоками?

Уроме того, вектора не порезаны, а лучше им быть распределёнными между процессами.

И я объяснял как делать матвек, и почему транспонированную матрицу хранить не нужно.

**Comment by 3Jlou4eJluk** [comment]
Насчёт распределения равномерного - вы говорили, что так можно, когда
объясняли на доске способы распределения. Насчёт не порезанных строк - это
уже блочная версия, правильно понимаю? Насчёт матвека - постараюсь
исправить.

**Comment by dzheltkov** [comment]
Это было для ILU, там это для балансировки распределения по памяти и вычислений, здесь так делать зачем?
Кроме того, даже если распределять так, такие ветвления не нужны.

Не строки не порезаны, а векторы в методе. Это и больше пересылок (при этом ещё и используется Allreduce, хотя если избавиться от ненужных ветвлений и считать на каждом процессе только короткий вектор, соответствующий той части матрицы, которую он хранит, то достаточно Allgather, а если разрезать векторы - то достаточно вообще неблокирующих циклических пересылок во время матвека) и больше памяти (матрица может быть разреженной или иметь другую структуру, и тогда непорезанные векторы могут занимать существенно больше порезанной матрицы).

То есть, в реальности для всех операций, кроме непосредственно матвека и вектора на который происходит умножение в матвеке (но это решается или Allgather для одного вектора в матвеке, и у вас тогда только один длинный вектор, или, лучше, неблокирующими циклическими пересылками, и тогда длинных векторов вообще нет, а пересылки маскируются за вычислениями), используется лишь часть вектора (только с нужными остатками от деления на число процессов в вашей реализации). 
Зачем хранить их целиком?

**Comment by dzheltkov** [comment]
Матрица всё ещё неудобно передаётся в функцию, реализующую метод. То есть у пользователя матрица может храниться совершенно иначе, лучше бы передавать или шаблонный класс, в котором обязан быть реализован необъходимый метод, или 'std::functional', или, если уж делать в C-стиле, 'void *' и указатели на функцию (или функцию), реализующую матвек.
Также не хватает отчёта и правил сборки CMake.
В остальном вроде более-менее, я могу поставить 4 за свою часть (уточняйте у Станислава Викторовича про питон).

**Comment by 3Jlou4eJluk** [comment]
Добрый день,

Я согласен на хор, правильно ли я понимаю, что мне нужно у Станислав
Викторовича сейчас оценку получить?


С уважением,
Калинин Михаил.

Ср, 20 сент. 2023 г. в 11:19, dzheltkov ***@***.***>:

**Comment by dzheltkov** [comment]
Нет, я говорил про свою часть.
Сейчас Станислав Викторович проверит часть по питону, узнаем, на какую оченку по своей части он согласен, выведем итоговую.

**Comment by 3Jlou4eJluk** [comment]
Да, именно это я и подразумевал.


С уважением,
Калинин Михаил.

ср, 20 сент. 2023 г., 14:31 dzheltkov ***@***.***>:
